src/main/java/bio/terra/service/load/flight/LoadUnlockStep.java

```diff
@@ -1,33 +1,78 @@
+import bio.terra.service.load.exception.LoadLockFailureException;
 import bio.terra.stairway.FlightContext;
 import bio.terra.stairway.StepResult;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Optional;
 
 public class LoadUnlockStep extends DefaultUndoStep {
   private final LoadService loadService;
+  private final String userSuppliedLockName;
 
   /**
-   * This step is meant to be shared by dataset and filesystem flights for unlocking the load tag.
+   * This step is meant to be shared by dataset and filesystem flights for removing {@code
+   * lockName}'s lock held on the load tag in the targeted dataset.
    *
    * <p>It expects the following to be available in the flight context:
    *
    * <ul>
-   *   <li>{@link LoadMapKeys#LOAD_TAG} in the input parameters or working map
+   *   <li>(Optional) {@link LoadMapKeys#LOAD_TAG} in the input parameters or working map (if
+   *       unavailable, any lock associated with the dataset and {@code lockName} will be cleared)
    *   <li>{@link JobMapKeys#DATASET_ID} in the input parameters
    * </ul>
+   *
+   * @param lockName the name of the lock to remove (if unspecified, defaults to this flight's ID)
    */
-  public LoadUnlockStep(LoadService loadService) {
+  public LoadUnlockStep(LoadService loadService, String lockName) {
     this.loadService = loadService;
+    this.userSuppliedLockName = lockName;
+  }
+
+  /**
+   * This step is meant to be shared by dataset and filesystem flights for removing this flight's
+   * lock held on the load tag in the targeted dataset.
+   *
+   * <p>It expects the following to be available in the flight context:
+   *
+   * <ul>
+   *   <li>(Optional) {@link LoadMapKeys#LOAD_TAG} in the input parameters or working map (if
+   *       unavailable, any lock associated with the dataset and this flight's ID will be cleared)
+   *   <li>{@link JobMapKeys#DATASET_ID} in the input parameters
+   * </ul>
+   */
+  public LoadUnlockStep(LoadService loadService) {
+    this(loadService, null);
+  }
+
+  @VisibleForTesting
+  public String getUserSuppliedLockName() {
+    return this.userSuppliedLockName;
   }
 
   @Override
   public StepResult doStep(FlightContext context) {
-    LoadLockKey loadLockKey = loadService.getLoadLockKey(context);
-    loadService.unlockLoad(loadLockKey, context.getFlightId());
+    String lockName = getLockName(context);
+    try {
+      LoadLockKey loadLockKey = loadService.getLoadLockKey(context);
+      loadService.unlockLoad(loadLockKey, lockName);
+    } catch (LoadLockFailureException ex) {
+      // The flight context may not have a load tag.  We'll clear any lock associated with the
+      // flight's dataset and lock name in that case.
+      loadService.unlockLoad(IngestUtils.getDatasetId(context), lockName);
+    }
     return StepResult.getStepResultSuccess();
   }
+
+  /**
+   * @return the name of the lock to remove (if unspecified, defaults to this flight's ID)
```

Comment:
I worry about the complexity of making steps work in different ways based on how they're used. Another approach would be to make this a required flight context property, but that would require adding an extra step to the flight to create the data this step needs.

Reply:
Good point -- now that this is in its own step, we have two simple steps rather than one complex one, while still reusing the service code.